hadoop-hdds/docs/content/design/storage-class.md

```diff
@@ -19,10 +19,331 @@ author: Marton Ele
+```
+
+This defines a replication scheme where only two replicas are enough from closed containers, and container will be erasure encoded under the hood if the 90% of the content is not used in the last 30 days.
+
+Please note that:
+
+ * All the low-level details of the replication rules can be configured here by the administrators
+ * Configuration is not global and not cluster-level, one can have different configuration for different storage-classes (which means for different keys/containers)
+ * Users dont' need to face with these details as they can use the storage-class (or just use the pre-created buckets and use default storage-class) abstraction
+
+## [DEV] Give flexibility to the developers
+
+Storage-class abstraction provides an easy way to plug in newer replication schemes. New type of replications (like EC) can be supported easily as the system will be prepared to allocate different type of containers.
+
+## [ADMIN] Better upgrade support
+
+Let's imagine that a new type of Open container replication is introduced (`RATIS-STREAM/THREE` instead of `RATIS/THREE`). If storage-classes are stored with the keys and containers instead of the direct replication rules we can:
+
+ 1. Easily change the replicaiton method of existing buckets/keys
+ 2. Turn on experimental features for specific buckets
+
+
+## [ADMIN] Change the cluster-wide replication
+
+An admin may decide to set a custom policy for an entire cluster.
+
+```
+ozone sh prefix setClass --storage-class=EC_6_3 /
+```
+
+# Unsupported use cases
+
+The following use cases are specifically unsupported.
+
+## [USER] Change the replication policy for a pre-existing key
+
+Changing the replication policy for a pre-existing key will require data movement and reauthoring containers and hence it is unsupported.
+
+## [USER] Defining storage-classes using Hadoop Compatible File System interface
+
+It's not possible to defined storage-class (or any replication rule) with using *Hadoop Compatible File System* interface. However storage-class defined on bucket level (or prefix level) will be inherited, even if the keys are created view the `o3fs://` or `o3s://` interfaces
+
+# The storage-class as an abstraction
+
+The previos section explained some user facing property of the storage-class concept. This section explains the concept compared to the existing Ozone design.
+
+## Containers in more details
+
+Container is the unit of replication of Ozone. One Container can store multiple blocks (default container size is 5GB) and they are replicated together. Datanodes report only the replication state of the Containers back to the Storage Container Manager (SCM) which makes it possible to scale up to billions of objects.
+
+The identifier of a block (BlockId) containers ContainerId and LocalId (ID inside the container). ContainerId can be used to find the right Datanode which stores the data. LocalId can be used to find the data inside one container.
+
+Container type defines the following:
```

Comment:
A container does not really define this today. There is a lot of replication specific logic encoded all over the system. If you want to introduce new replication types, all of that must be refactored and made generic behind interfaces. We are talking about months of refactoring/rewrite effort.

I think if EC is the immediate use case, it is easier to introduce a new EC-enabled write path.

Reply:
I made it more clear. We have replication mechanism which uses different type of containers. Today we have one replication mechanism which uses open and closed containers and defines the transitions between them.

I agree that a new replication mechanism requires code changes.

But *choosing* and *configuring* the replication for specific containers / keys can be done easily with this abstraction.